src/main/kotlin/com/team4099/robot2023/util/FieldFrameEstimator.kt

```diff
@@ -0,0 +1,224 @@
+package com.team4099.robot2023.util
+
+import com.team4099.lib.hal.Clock
+import com.team4099.lib.math.asTransform2d
+import edu.wpi.first.math.Matrix
+import edu.wpi.first.math.Nat
+import edu.wpi.first.math.VecBuilder
+import edu.wpi.first.math.numbers.N1
+import edu.wpi.first.math.numbers.N3
+import org.littletonrobotics.junction.Logger
+import org.team4099.lib.geometry.Pose2d
+import org.team4099.lib.geometry.Pose2dWPILIB
+import org.team4099.lib.geometry.Transform2d
+import org.team4099.lib.geometry.Translation2d
+import org.team4099.lib.geometry.Twist2d
+import org.team4099.lib.units.base.Time
+import org.team4099.lib.units.base.inMeters
+import org.team4099.lib.units.base.inSeconds
+import org.team4099.lib.units.base.meters
+import org.team4099.lib.units.base.seconds
+import org.team4099.lib.units.derived.inRadians
+import org.team4099.lib.units.derived.radians
+import java.util.NavigableMap
+import java.util.TreeMap
+import kotlin.Comparator
+import kotlin.collections.ArrayList
+
+class FieldFrameEstimator(stateStdDevs: Matrix<N3?, N1?>) {
+  // Maintains the state of the field frame transform for the period of time before the currently
+  // tracked history
+  private var baseOdometryTField: Transform2d =
+    Transform2d(Translation2d(0.meters, 0.meters), 0.radians)
+
+  // Maintains the latest state of the field frame transform, including the currently tracked
+  // history
+  private var odometryTField: Transform2d =
+    Transform2d(Translation2d(0.meters, 0.meters), 0.radians)
+
+  private val updates: NavigableMap<Time, PoseUpdate> = TreeMap()
+  private val q: Matrix<N3?, N1?> = Matrix(Nat.N3(), Nat.N1())
+
+  /** Returns the latest robot pose based on drive and vision data. */
+  fun getLatestOdometryTField(): Transform2d {
+    return odometryTField
+  }
+
+  /** Resets the field frame transform to a known pose. */
+  fun resetFieldFrameFilter(transform: Transform2d) {
+    baseOdometryTField = transform
+    updates.clear()
+    update()
+  }
+
+  /** Records a new drive movement. */
+  fun addDriveData(timestamp: Time, odomTRobot: Pose2d) {
```

Comment:
i think it's worth adding a `SwerveDriveOdometry` here you can take these twists and also keep track of the `odoTRobot` pose independently (for when we want to do path following stuff) and have another method that applies the transform `odoTField` to `odoTRobot` and resets `SwerveDriveOdometry`

Reply:
i would prefer to do this here rather than dt for the sole reason of consolidating the objects you feed drive twists into as a vision measurement, also it makes sense to be able to access `odoTRobot` in ur pose estimator

Reply:
I would like to keep odometry separate from the field frame estimator -- the field frame estimator should not care about how odometry is calculated etc IMO. I don't think you need `odomTRobot` anywhere in this estimator outside of drive data updates anyway. Basically there are then two separate pose estimators, one for the odometry frame and one for the field frame

Reply:
There's no guarantee that the drive data this loop cycle will be incorporated into the vision frame's calculation if you don't put it here though. It really depends what gets called first, addDriveData or addVisionMeasurement, the answer to which depends on the order on which subsystem periodic's are called? Idk if the results would be meaningfully different bc it's one loop cycle but das why I wanna put it here

Reply:
I mean this is equivalent to the older pose estimator and that worked? I am inclined to think it will be fine. If drivetrain periodic is called after vision periodic anyway the odometry won't be updated because it won't have swerve module states from that timestamp